File: problems/99/solution.go
package main
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func recoverTree(root *TreeNode)  {
    var traverse func(*TreeNode)
    var first, last, prev *TreeNode
    
    traverse = func (node *TreeNode) {
        if node == nil {
            return
        }
        
        traverse(node.Left)
        
        if prev != nil && prev.Val > node.Val {
            if first == nil {
                first = prev
            }
            
            last = node
        }
        
        prev = node
        
        traverse(node.Right)
    }
    
    traverse(root)
    
    first.Val, last.Val = last.Val, first.Val
}

File: problems/99/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode *one, *two, *prev;
    
    void recoverTree(TreeNode* root) {
        traverse(root);
        
        one->val ^= two->val;
        two->val ^= one->val;
        one->val ^= two->val;
    }
    
    void traverse(TreeNode* node) {
        if (!node) {
            return;
        }
        
        traverse(node->left);
        
        if (prev && prev->val > node->val) {
            if (!one) {
                one = prev;
            }
            
            two = node;
        }
        
        prev = node;
        traverse(node->right);
    }
};